pkg/sql/sem/tree/expr.go

```diff
@@ -1624,7 +1624,15 @@ type IndirectionExpr struct {
 
 // Format implements the NodeFormatter interface.
 func (node *IndirectionExpr) Format(ctx *FmtCtx) {
-	exprFmtWithParen(ctx, node.Expr)
+	// If the subExpr is a CastExpr, we need to wrap it in a ParenExpr,
+	// otherwise the indirection will get interpreted as part of the type.
+	// Ex. ('{a}'::_typ)[1] vs. '{a}'::_typ[1].
+	if _, isCast := node.Expr.(*CastExpr); isCast {
+		withParens := ParenExpr{Expr: node.Expr}
+		exprFmtWithParen(ctx, &withParens)
```

Comment:
interesting, my preference would have been to always wrap CastExpr with parens in `exprFmtWithParen` but too bad that seems to break things...

i don't know if i'm ready to advocate for this yet but: what if we _always_ wrap the inner expr of an IndirectionExpr with parens...? just curious how noisy that gets and if there are major downsides